pkg/controllers/user/helm/common/extra_args_injection.go

```diff
@@ -0,0 +1,60 @@
+package common
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/rancher/rancher/pkg/ref"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+type InjectAppArgsFunc func(obj *v3.App) (content map[string]string)
+
+const (
+	systemCatalogName = "system-library"
+)
+
+var (
+	extraArgsFuncs = []InjectAppArgsFunc{
+		injectDefaultRegistry,
+		injectClusterInfo,
+	}
+)
+
+func injectDefaultRegistry(obj *v3.App) map[string]string {
+	values, err := url.Parse(obj.Spec.ExternalID)
+	if err != nil {
+		logrus.Errorf("check catalog type failed: %s", err.Error())
+	}
+
+	catalogWithNamespace := values.Query().Get("catalog")
+	split := strings.SplitN(catalogWithNamespace, "/", 2)
+	catalog := split[len(split)-1]
+
+	reg := settings.SystemDefaultRegistry.Get()
+	if catalog != systemCatalogName || reg == "" {
+		return nil
+	}
+	return map[string]string{"systemDefaultRegistry": reg}
+}
+
+func injectClusterInfo(obj *v3.App) map[string]string {
+	clusterName, projectName := ref.Parse(obj.Spec.ProjectName)
+	return map[string]string{
+		"clusterName": clusterName,
+		"projectName": projectName,
+	}
+}
+
+func GetExtraArgs(app *v3.App) map[string]string {
```

Comment:
This implies that `global.clusterName` & `global.projectName` become preserved keys for all apps. Is it better if we set them when deploying monitoring apps? I'm not sure if they are useful for other apps though.

Reply:
It is useful and it is our responsibility to let app know which cluster and project it's in.